core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1490,6 +1495,39 @@ class Log(@volatile var dir: File,
     }
   }
 
+  private def recordConsumerFetchLagMetrics(segmentBaseOffset: Long, segmentSizeInBytes: Long, 
+                                    endOffset: Long, fetchInfo: FetchDataInfo){   
+    // record fetch time lag in milliseconds 
+    // time lag = activeSegment.maxTimestampSoFar - MaximumTimestamp of first batch fetched for the startOffset  
+    val iterator = fetchInfo.records.batches.iterator
+    if (iterator.hasNext()){
+      val batch = iterator.next()
+      if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
+        val startOffsetTimestamp = batch.maxTimestamp()
+        val maxTimestampSoFar = activeSegment.maxTimestampSoFar
+        if (startOffsetTimestamp != -1 && maxTimestampSoFar != -1) {
+          val fetchLagInMS = maxTimestampSoFar - startOffsetTimestamp
```

Comment:
I think this is a reasonable way to achieve this, but I think it could jump around a lot. I would expect that sometimes the lag would increase just because the next record batch read had higher timestamps. This might be OK overall, but I believe you could still see lag increase even if no new data is written.

I also wonder whether you will ever see lag drop to 0, because you are only updating the lag via the first record batch.

Reply:
Yes I agree, we really care about the distribution

Reply:
We have to benchmark this to make sure it doesn't add measurable overhead on the read path.